src/app/fddev/quic_trace/fd_quic_trace_main.c

```diff
@@ -0,0 +1,111 @@
+}
+
+void
+quic_trace_cmd_fn( args_t *         args,
+                   config_t * const config ) {
+  (void)args;
+
+  fd_topo_t * topo = &config->topo;
+  fd_topo_join_workspaces( topo, FD_SHMEM_JOIN_MODE_READ_ONLY );
+  fd_topo_fill( topo );
+
+  fd_topo_tile_t * quic_tile = NULL;
+  for( ulong tile_idx=0UL; tile_idx < topo->tile_cnt; tile_idx++ ) {
+    if( 0==strcmp( topo->tiles[ tile_idx ].name, "quic" ) ) {
+      quic_tile = &topo->tiles[ tile_idx ];
+      break;
+    }
+  }
+  if( !quic_tile ) FD_LOG_ERR(( "QUIC tile not found in topology" ));
+
+  /* Ugly: fd_quic_ctx_t uses non-relocatable object addressing.
+     We need to rebase pointers.  foreign_{...} refer to the original
+     objects in shared memory, local_{...} refer to translated copies. */
+
+  void *                quic_tile_base   = fd_topo_obj_laddr( topo, quic_tile->tile_obj_id );
+  fd_quic_ctx_t const * foreign_quic_ctx = quic_tile_base;
+  FD_TEST( foreign_quic_ctx->magic == FD_QUIC_TILE_CTX_MAGIC );
+  fd_quic_ctx_t * quic_ctx = &fd_quic_trace_ctx;
+  *quic_ctx                = *foreign_quic_ctx;
+  fd_quic_trace_ctx_remote =  foreign_quic_ctx;
+  FD_LOG_INFO(( "fd_quic_tile state at %p in tile address space", foreign_quic_ctx->self ));
+  FD_LOG_INFO(( "fd_quic_tile state at %p in local address space", quic_tile_base ));
+
+  quic_ctx->reasm = (void *)( (ulong)quic_tile_base + (ulong)quic_ctx->reasm - (ulong)quic_ctx->self );
+  quic_ctx->stem  = (void *)( (ulong)quic_tile_base + (ulong)quic_ctx->stem  - (ulong)quic_ctx->self );
+  quic_ctx->quic  = (void *)( (ulong)quic_tile_base + (ulong)quic_ctx->quic  - (ulong)quic_ctx->self );
+
+  fd_topo_link_t * net_quic = &topo->links[ quic_tile->in_link_id[ 0 ] ];
+  quic_ctx->in_mem = topo->workspaces[ topo->objs[ net_quic->dcache_obj_id ].wksp_id ].wksp;
+  FD_LOG_INFO(( "net->quic link at %p", (void *)quic_ctx->in_mem ));
+
+  /* Join shared memory objects
+     Mostly nops but verifies object magic numbers to ensure that
+     derived pointers are correct. */
+
+  FD_LOG_INFO(( "Joining fd_quic" ));
+  fd_quic_t * quic = fd_quic_join( quic_ctx->quic );
+  if( !quic ) FD_LOG_ERR(( "Failed to join fd_quic" ));
+
+  /* Locate original fseq objects
+     These are monitored to ensure the trace RX tile doesn't skip ahead
+     of the quic tile. */
+  fd_quic_trace_target_fseq = malloc( quic_tile->in_cnt * sizeof(ulong) );
+  for( ulong i=0UL; i<quic_tile->in_cnt; i++ ) {
+    fd_quic_trace_target_fseq[ i ] = quic_tile->in_link_fseq[ i ];
+  }
+
+  /* Redirect metadata writes to dummy buffers.
+     Without this hack, stem_run would attempt to write metadata updates
+     into the target topology which is read-only. */
+
+  /* ... redirect metric updates */
+  ulong * metrics = aligned_alloc( FD_METRICS_ALIGN, FD_METRICS_FOOTPRINT( quic_tile->in_cnt, quic_tile->out_cnt ) );
+  if( !metrics ) FD_LOG_ERR(( "out of memory" ));
+  fd_metrics_register( metrics );
+
+  /* ... redirect fseq updates */
```

Comment:
I don't think this does anything? The QUIC tile has already. booted once you do this, so wouldn't read these fseq pointers again?

Reply:
Oh lol I see how this works ... kind of an interesting strategy to just pretend it's a QUIC tile, but I guess works for now.

Reply:
@mmcgee-jump I don't see any other way. We need to consume net frags but we also need decryption keys. 
If we skip ahead of the real QUIC tile then the packets we see might not have encryption keys available.

Reply:
Oh I thought your comment was about something else. 
Eventually, I want to make it less hacky and make it a real tile. But I wasn't going to bother with these fdctl APIs just for a debugging tool